### PR TITLE
aws_ssm connection: add support for SSM document

### PR DESCRIPTION
**Pull request by blinkseb**
##### SUMMARY

This PR adds support for SSM document to the SSM connection plugin.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

community.aws.aws_ssm

##### ADDITIONAL INFORMATION

The new document parameters is directly forwarded to the SSM [`start_session`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ssm.html#SSM.Client.start_session) method.

Usage:
```
- name: Install a Nginx Package
  vars:
    ansible_connection: aws_ssm
    ansible_aws_ssm_bucket_name: nameofthebucket
    ansible_aws_ssm_region: us-west-2
    ansible_aws_ssm_document: nameofthecustomdocument
  tasks:
    - name: Install a Nginx Package
      yum:
        name: nginx
        state: present
```

### CODE REVIEW

File: plugins/connection/aws_ssm.py [hunk 1]
```diff
@@ -82,6 +82,10 @@
     version_added: 2.2.0
     vars:
     - name: ansible_aws_ssm_bucket_sse_kms_key_id
+  ssm_document:
+    description: SSM document to use when connecting.
+    vars:
+    - name: ansible_aws_ssm_document
```
Comment (alinabuzachis) [inline plugins/connection/aws_ssm.py:88]: @blinkseb Thank you for taking time to work on this feature idea. Please add `version_added: 3.1.0`

File: plugins/connection/aws_ssm.py [hunk 2]
```diff
@@ -343,7 +360,11 @@ def start_session(self):
         ssm_parameters = dict()
         client = self._get_boto_client('ssm', region_name=region_name, profile_name=profile_name)
         self._client = client
-        response = client.start_session(Target=self.instance_id, Parameters=ssm_parameters)
+        start_session_args = dict(Target=self.instance_id, Parameters=ssm_parameters)
+        document_name = self.get_option('ssm_document')
+        if document_name is not None:
+            start_session_args['DocumentName'] = document_name
+        response = client.start_session(**start_session_args)
```
Comment (alinabuzachis) [inline plugins/connection/aws_ssm.py:367]: Could you please add an integration test for this feature idea?

**Review by alinabuzachis — CHANGES_REQUESTED** [review]
Could you also please add a changelog fragment?

File: plugins/connection/aws_ssm.py (hunk 1 above)
Comment (blinkseb) [reply plugins/connection/aws_ssm.py:88]: done :)

File: plugins/connection/aws_ssm.py (hunk 2 above)
Comment (blinkseb) [reply plugins/connection/aws_ssm.py:368]: Sure, I tried something, let's see if it's working. Ideally, we'd create a custom SSM document, but unfortunately I can't find an ansible module for that, so instead, I just put the default SSM document used behind the scene by the SSM client

**Comment by blinkseb** [comment]
Thanks for your review! I added the changelog fragment and tried something for the integration tests

**Comment by marknet15** [comment]
@blinkseb It looks like the integration tests are currently disabled:
https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/connection_aws_ssm/aliases

I'm not sure of the history if the past issues are resolved or not etc, but you can try running them locally to verify all is working as expected you would need to specify the `--allow-disabled` flag so for example:
```
ansible-test integration --docker default -vvv connection_aws_ssm --allow-disabled
```

**Comment by blinkseb** [comment]
ok I'm progressing. I re-enabled the integration tests, like it was done in https://github.com/ansible-collections/community.aws/pull/763. I'm currently stuck with a permission error to create a new SSM document, so I opened a PR to add the necessary permissions here: https://github.com/mattclay/aws-terminator/pull/192

**Comment by markuman** [comment]
I run the tests locally. I _guess_ it all passes.  
But at the end, the test still return exit code 1. No idea why.


```
....
localhost                  : ok=16   changed=6    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   

AWS ACTIONS: ['ec2:DescribeInstances', 'ec2:TerminateInstances', 'iam:DeleteRole', 'iam:DetachRolePolicy', 'iam:GetRole', 'iam:ListAttachedRolePolicies', 'iam:ListInstanceProfilesForRole', 'iam:RemoveRoleFromInstanceProfile', 'kms:DescribeKey', 'kms:GetKeyPolicy', 'kms:GetKeyRotationStatus', 'kms:ListAliases', 'kms:ListGrants', 'kms:ListKeyPolicies', 'kms:ListResourceTags', 'kms:ScheduleKeyDeletion']
Command exited with status 1 after 567.5553092956543 seconds.
Run command: /usr/bin/python /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.01774883270263672 seconds.
Run command: /usr/bin/python2 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python2
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.01106405258178711 seconds.
Run command: /usr/bin/python2.6 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python2.6
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.23974013328552246 seconds.
Run command: /usr/bin/python2.7 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python2.7
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.01092529296875 seconds.
Run command: /usr/bin/python3 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python3
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.016592025756835938 seconds.
Run command: /usr/bin/python3.5 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python3.5
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.0272824764251709 seconds.
Run command: /usr/bin/python3.6 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python3.6
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.016504526138305664 seconds.
Run command: /usr/bin/python3.7 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python3.7
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.014443159103393555 seconds.
Run command: /usr/bin/python3.8 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python3.8
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.01379251480102539 seconds.
Run command: /usr/bin/python3.9 /root/ansible/test/lib/ansible_test/_data/versions.py
Working directory: /root/ansible_collections/community/aws
Program found: /usr/bin/python3.9
HOME=/root
LC_ALL=en_US.UTF-8
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
Command exited with status 0 after 0.015276432037353516 seconds.
NOTICE: To resume at this test target, use the option: --start-at connection_aws_ssm
ERROR: Command "./runme.sh -vvvv -e '@/root/ansible_collections/community/aws/tests/output/.tmp/integration/connection_aws_ssm-9b3389a1-ÅÑŚÌβŁÈ/tests/integration/config-file-tpixtwm6.json'" returned exit status 1.
Cleaning up temporary python directory: /tmp/python-73fse3in-ansible
Command exited with status 1 after 608.0626201629639 seconds.
Run command: docker exec 69a0350a43534dd4965c88f01d5c86423716ab768a1e32ebf9f5dcae60ce5fba tar czf /root/output.tgz --exclude .tmp -C /root/ansible_collections/community/aws/tests output
Working directory: /home/m/git/ansible_collections/community/aws
Program found: /usr/bin/docker
HOME=/home/m
LC_ALL=en_US.UTF-8
PATH=/home/m/.local/bin:/usr/local/bin:/usr/bin:/var/lib/snapd/snap/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/m/.local/bin/
SSH_AUTH_SOCK=/run/user/1000/keyring/ssh
Command exited with status 0 after 0.08580565452575684 seconds.
Run command: docker exec -i 69a0350a43534dd4965c88f01d5c86423716ab768a1e32ebf9f5dcae60ce5fba dd if=/root/output.tgz bs=65536
Working directory: /home/m/git/ansible_collections/community/aws
Program found: /usr/bin/docker
HOME=/home/m
LC_ALL=en_US.UTF-8
PATH=/home/m/.local/bin:/usr/local/bin:/usr/bin:/var/lib/snapd/snap/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/m/.local/bin/
SSH_AUTH_SOCK=/run/user/1000/keyring/ssh
Command exited with status 0 after 0.08487987518310547 seconds.
Run command: tar oxzf /tmp/ansible-result-h_w2h_mw.tgz -C /home/m/git/ansible_collections/community/aws/tests
Working directory: /home/m/git/ansible_collections/community/aws
Program found: /usr/bin/tar
HOME=/home/m
LC_ALL=en_US.UTF-8
PATH=/home/m/.local/bin:/usr/local/bin:/usr/bin:/var/lib/snapd/snap/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/m/.local/bin/
SSH_AUTH_SOCK=/run/user/1000/keyring/ssh
Command exited with status 0 after 0.006506204605102539 seconds.
Run command: docker rm -f 69a0350a43534dd4965c88f01d5c86423716ab768a1e32ebf9f5dcae60ce5fba
Working directory: /home/m/git/ansible_collections/community/aws
Program found: /usr/bin/docker
HOME=/home/m
LC_ALL=en_US.UTF-8
PATH=/home/m/.local/bin:/usr/local/bin:/usr/bin:/var/lib/snapd/snap/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/m/.local/bin/
SSH_AUTH_SOCK=/run/user/1000/keyring/ssh
Command exited with status 0 after 0.5264806747436523 seconds.
ERROR: Command "docker exec 69a0350a43534dd4965c88f01d5c86423716ab768a1e32ebf9f5dcae60ce5fba /usr/bin/env ANSIBLE_TEST_CONTENT_ROOT=/root/ansible_collections/community/aws LC_ALL=en_US.UTF-8 /usr/bin/python3.8 /roo
```

**Comment by markuman** [comment]
recheck

File: plugins/connection/aws_ssm.py [hunk 3]
```diff
@@ -82,6 +82,11 @@
     version_added: 2.2.0
     vars:
     - name: ansible_aws_ssm_bucket_sse_kms_key_id
+  ssm_document:
+    description: SSM document to use when connecting.
+    vars:
+    - name: ansible_aws_ssm_document
+    version_added: 3.1.0
```
Comment (markuman) [inline plugins/connection/aws_ssm.py:89]:
```suggestion
    version_added: 3.2.0
```

**Comment by markuman** [comment]
Ah I think the terminator needs more permission to run this integration test

```
"An error occurred (AccessDeniedException) when calling the DeleteDocument operation: 

User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=remote=zuul-cloud is not authorized to perform:

ssm:DeleteDocument on resource: arn:aws:ssm:us-east-1:966509639900:document/ansible-custom-document because no identity-based policy allows the ssm:DeleteDocument action"
```

I made a PR for that https://github.com/mattclay/aws-terminator/pull/201

**Comment by blinkseb** [comment]
Helle everyone, is there something additional I need to do to have this PR merged? Thanks!

File: plugins/connection/aws_ssm.py [hunk 4]
```diff
@@ -82,6 +82,11 @@
     version_added: 2.2.0
     vars:
     - name: ansible_aws_ssm_bucket_sse_kms_key_id
+  ssm_document:
+    description: SSM document to use when connecting.
+    vars:
+    - name: ansible_aws_ssm_document
+    version_added: 3.2.0
```
Comment (markuman) [inline plugins/connection/aws_ssm.py:89]:
```suggestion
    version_added: 5.1.0
```

**Review by alinabuzachis — APPROVED** [review]

**Comment by alinabuzachis** [comment]
recheck

**Comment by markuman** [comment]
Okey, we still have CI permissions

>  User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=remote=zuul-cloud is not authorized to perform: ssm:DeleteDocument on resource: arn:aws:ssm:us-east-1:966509639900:document/ansible-custom-document because no identity-based policy allows the ssm:DeleteDocument action",

I think someone must review and merge https://github.com/mattclay/aws-terminator/pull/192 before we can get forward here.

**Comment by gravesm** [comment]
recheck

**Comment by markuman** [comment]
recheck

**Comment by markuman** [comment]
hmmm

>     "msg": "timed out waiting for ping module test: An error occurred (AccessDenied) when calling the PutObject operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=remote=zuul-cloud is not authorized to perform: kms:GenerateDataKey on resource: arn:aws:kms:us-east-1:966509639900:key/5aa12287-6092-4fcc-9e8a-a08805e2706c because no identity-based policy allows the kms:GenerateDataKey action"

cc @gravesm

**Comment by markuman** [comment]
@blinkseb do you have some time to address the missing permission in a PR again?

**Review by tremble — APPROVED** [review]
Code looks good.  I'm reworking the tests a little so that we can test things still work *without* an SSM document

**Comment by blinkseb** [comment]
> Code looks good. I'm reworking the tests a little so that we can test things still work _without_ an SSM document

thanks a lot!

**Comment by tremble** [comment]
@blinkseb Thanks for taking the time to submit this PR, I'm sorry it took quite so long to get everything working.  With integration tests now in place hopefully we can start to clear some of the backlog around the SSM connection plugin.

**Pull request merged**